tools/clang/unittests/HLSLExec/ExecutionTest.cpp

```diff
@@ -11096,6 +11135,402 @@ TEST_F(ExecutionTest, PackUnpackTest) {
+    // the test. We use it to override the values set on the st:ShaderOp from
+    // ShaderOpArith.xml. We update with values defined in ShadoerOpArithTable.xml.
+    [&](LPCSTR Name, std::vector<BYTE> &Data, st::ShaderOp *pShaderOp) {
+      LogCommentFmt(L"RunShaderOpTest CB. Shader Name: %S", Name);
+      VERIFY_IS_TRUE(0 == _stricmp(Name, "SLongVectorBinaryOp<T>"));
+
+      const size_t size = sizeof(SLongVectorBinaryOp<T, N>) * inputVectorCount;
+      Data.resize(size);
+
+      // Treat the data as an array of SLongVectorBinaryOp<T N> structs.
+      // One struct for each set of <value> tags in the  ShaderOpArithTable.xml
+      // A set of value tags is the corresponding <value> tags across <Parameters>
+      SLongVectorBinaryOp<T, N> *pPrimitives = (SLongVectorBinaryOp<T, N> *)Data.data();
+
+      for (size_t i = 0; i < inputVectorCount ; ++i) {
+        SLongVectorBinaryOp<T, N> *p = &pPrimitives[i];
+
+        std::copy(vecInput1[i].begin(), vecInput1[i].end(), p->vecInput1.begin());
+
+        if(!vecInput2.empty())
+        {
+          std::copy(vecInput2[i].begin(), vecInput2[i].end(), p->vecInput2.begin());
+        }
+        else if(!scalarInputs.empty())
+        {
+          VERIFY_IS_TRUE(scalarInputs[i].size() == 1, L"Scalar input vector must have exactly one element.");
+          p->scalarInput = scalarInputs[i][0];
+        }
+      }
+
+      // Override the shader name and text (source hlsl) with the values from ShaderOpArithTable.xml
+      pShaderOp->Shaders.at(0).Target = Target.m_psz;
+      pShaderOp->Shaders.at(0).Text = Text.m_psz;
+
+      // And override the compiler arguments. We pass in the element
+      // type for the vectors and their size.
+      pShaderOp->Shaders.at(0).Arguments = compilerOptions;
+    }
+  );
+
+  // Map the data from GPU to CPU memory so we can verify our expectations.
+  MappedData data;
+  test->Test->GetReadBackData("SLongVectorBinaryOp<T>", &data);
+
+  // Cast the buffer back into an array of the structs we expect.
+  SLongVectorBinaryOp<T, N> *pPrimitive = (SLongVectorBinaryOp<T, N>*)data.data();
+  for(size_t i = 0; i < inputVectorCount; i++)
+  {
+    SLongVectorBinaryOp<T, N> *p = &pPrimitive[i];
+    std::vector<T> vecOutput(p->vecOutput.begin(), p->vecOutput.end());
+
+    VERIFY_SUCCEEDED(DoVectorsMatch<T>(vecOutput, vecExpected[i], Validation_Tolerance));
+  }
+}
+
+template <typename T, std::size_t N = 4,
+typename = std::enable_if_t<std::is_arithmetic_v<T>>>
+struct SLongVectorUnaryOp {
+  T clampArgC;
+  T clampArgT;
+  std::array<T, N> vecInput1;
+  std::array<T, N> vecOutput;
+};
+
+TEST_F(ExecutionTest, LongVector_UnaryOpTest) {
+  WEX::TestExecution::SetVerifyOutput verifySettings(
+      WEX::TestExecution::VerifyOutputSettings::LogOnlyFailures);
+    LongVectorUnaryOpTestBase<float, 4>();
+}
+
+template <typename T, std::size_t N>
+//typename = std::enable_if_t<std::is_arithmetic_v<T>>>
+void ExecutionTest::LongVectorUnaryOpTestBase() {
```

Comment:
This code is extremely similar to BinaryTestOpBase, consider factoring out common logic into helpers.